gcc/rust/privacy/rust-reachability.cc

```diff
@@ -31,6 +31,47 @@ maybe_get_vis_item (std::unique_ptr<HIR::Item> &item)
   return static_cast<HIR::VisItem *> (item.get ());
 }
 
+ReachLevel
+ReachabilityVisitor::initial_reach (const HIR::Visibility &item_visibility)
+{
+  if (item_visibility.is_public ())
+    return ReachLevel::Reachable;
+
+  return ReachLevel::Unreachable;
+}
+
+void
+ReachabilityVisitor::visit_generic_predicates (
+  const std::vector<std::unique_ptr<HIR::GenericParam>> &generics,
+  ReachLevel item_reach)
+{
+  if (item_reach == ReachLevel::Unreachable)
+    return;
+
+  for (auto &generic : generics)
+    {
+      if (generic->get_kind () == HIR::GenericParam::TYPE)
+	{
+	  TyTy::BaseType *generic_ty = nullptr;
+	  rust_assert (
+	    ty_ctx.lookup_type (generic->get_mappings ().get_hirid (),
+				&generic_ty));
+
+	  // FIXME: Can we really get anything else than a TyTy::PARAM here?
+	  // Should we change this to an assertion instead?
+	  if (generic_ty->get_kind () == TyTy::PARAM)
```

Comment:
yeah thats that right, this might as well be an assertion, I should really update that in the rest of the type resolution steps.